File: src/finance_analysis/interfaces/api/v1/schemas/market_sentiment.py
"""Authenticated observation contracts. Ratios are decimal; money is CNY."""

from datetime import date, datetime
from typing import Any, Literal
from pydantic import BaseModel, Field, model_validator

PoolKind = Literal["limit_up", "limit_down", "limit_break"]


class Promotion(BaseModel):
    source_date: date | None
    target_date: date
    complete: bool
    numerator: int | None
    denominator: int | None
    ratio: float | None
    promoted_codes: list[str]
    not_promoted_codes: list[str]


class Observation(BaseModel):
    trade_date: date
    previous_trade_date: date | None
    scope: str
    rule_version: str
    early_time_threshold: str
    state: Literal["UNKNOWN", "ICE", "COOLING", "REPAIR", "DIVERGENCE", "CLIMAX", "ACTIVE", "NEUTRAL"]
    heat_score: float | None
    state_reasons: list[str]
    upstream_total: int
    excluded_st_count: int
    excluded_new_count: int
    excluded_union_count: int
    unknown_scope_count: int
    scope_complete: bool
    boards_complete: bool
    limit_up_count: int | None
    first_board_count: int | None
    multi_board_count: int | None
    highest_board: int | None
    board_distribution: dict[str, int | None]
    unconfirmed_board_count: int
    early_limit_up_count: int | None
    valid_limit_up_time_count: int | None
    time_coverage: float | None
    early_limit_up_ratio: float | None
    seal_retention_median: float | None
    valid_seal_retention_count: int | None
    seal_retention_coverage: float | None
    seal_money_sum: float | None
    valid_seal_money_count: int | None
    seal_money_coverage: float | None
    reasons: list[dict[str, Any]]
    promotions: dict[str, Promotion]
    changes: dict[str, float | None]
    quality: dict[str, Any]
    supplements: dict[str, Any]
    source_timestamp: datetime
    fetched_at: datetime
    generated_at: datetime


class OverviewResponse(BaseModel):
    trade_date: date | None
    expected_trade_date: date
    observation: Observation | None
    industry_top: list[dict[str, Any]]


class HistoryResponse(BaseModel):
    dates: list[date]
    items: list[Observation | None]


class PoolResponse(BaseModel):
    trade_date: date | None
    kind: PoolKind
    available: bool
    total: int | None
    upstream_total: int | None
    page: int
    size: int
    items: list[dict[str, Any]]
    basis: str


class RunRequest(BaseModel):
    trade_date: date | None = None
    backfill_days: int | None = Field(None, ge=1, le=31)
    missing_only: bool = True

    @model_validator(mode="after")
    def exclusive(self):
        if self.trade_date is not None and self.backfill_days is not None:
            raise ValueError("trade_date 与 backfill_days 不能同时提供")
        return self


class RunResponse(BaseModel):
    task_id: str
    status: str = "pending"


class LadderSource(BaseModel):
    trade_date: date
    requested_trade_date: date | None
    source_kind: str
    source_timestamp: datetime
    fetched_at: datetime
    total: int
    item_count: int
    quality: dict[str, Any]
    window: dict[str, Any]
    items: list[dict[str, Any]]


class LadderResponse(BaseModel):
    as_of: date | None
    source: LadderSource | None
    basis: str

File: web/src/api/marketSentiment.ts
import apiClient from './index';
import { toCamelCase } from './utils';
export interface Promotion {
  sourceDate: string | null; targetDate: string; complete: boolean;
  numerator: number | null; denominator: number | null; ratio: number | null;
  promotedCodes: string[]; notPromotedCodes: string[];
}
export interface Observation {
  tradeDate: string; previousTradeDate: string | null; ruleVersion: string; scope: string;
  earlyTimeThreshold: string; state: string; heatScore: number | null; stateReasons: string[];
  upstreamTotal: number; excludedStCount: number; excludedNewCount: number; excludedUnionCount: number;
  unknownScopeCount: number; scopeComplete: boolean; boardsComplete: boolean;
  limitUpCount: number | null; firstBoardCount: number | null; multiBoardCount: number | null; highestBoard: number | null;
  boardDistribution: Record<string, number | null>; unconfirmedBoardCount: number;
  earlyLimitUpCount: number | null; validLimitUpTimeCount: number | null; timeCoverage: number | null;
  earlyLimitUpRatio: number | null; sealRetentionMedian: number | null;
  validSealRetentionCount: number | null; sealRetentionCoverage: number | null;
  sealMoneySum: number | null; validSealMoneyCount: number | null; sealMoneyCoverage: number | null;
  reasons: { reason: string; count: number }[]; promotions: Record<string, Promotion>;
  changes: Record<string, number | null>; quality: { optionalErrors?: Record<string, string> };
  supplements: Record<string, { total: number }>;
  sourceTimestamp: string; fetchedAt: string; generatedAt: string;
}
export interface Overview {
  tradeDate: string | null; expectedTradeDate: string; observation: Observation | null;
  industryTop: { tradeDate: string; industryCode: string; industryName: string; strengthScore: number; state: string }[];
}
export interface History { dates: string[]; items: (Observation | null)[] }
export interface PoolRow {
  thscode: string; name: string; isSt?: boolean | null; isNew?: boolean | null;
  inScope?: boolean | null; continueDayText?: string | null; consecutiveBoards?: number | null;
  priceChangeRatio?: number | null; priceChangeRatioPct?: number | null; limitUpTime?: string | null;
  sealMoney?: number | null; maxSealMoney?: number | null; sealRetention?: number | null;
  limitUpReason?: string | null; qualityIssues?: string[];
  firstLimitTime?: string | null; lastLimitTime?: string | null; openTimes?: number | null;
}
export interface Pool { tradeDate: string | null; kind: string; available: boolean; total: number | null;
  upstreamTotal: number | null; page: number; size: number; items: PoolRow[]; basis: string }
export interface Ladder {
  source: null | { tradeDate: string; sourceTimestamp: string; fetchedAt: string;
    window: { length: number; dateList: string[]; boardCaps: Record<string, number> };
    items: { date: string; boards: Record<string, { thscode: string; name: string; boardNum: number }[]> }[] };
}
const base = '/api/v1/market-sentiment';
export const marketSentimentApi = {
  async overview(tradeDate?: string): Promise<Overview> {
    return toCamelCase((await apiClient.get(`${base}/overview`, { params: { trade_date: tradeDate } })).data);
  },
  async history(endDate?: string): Promise<History> {
    return toCamelCase((await apiClient.get(`${base}/history`, { params: { end_date: endDate, days: 30 } })).data);
  },
  async dates(): Promise<string[]> { return (await apiClient.get(`${base}/dates`)).data; },
  async pool(params: { trade_date?: string; kind?: string; board?: string; q?: string; reason?: string; scope?: string; page?: number }): Promise<Pool> {
    return toCamelCase((await apiClient.get(`${base}/pool`, { params: { ...params, size: 50 } })).data);
  },
  async ladder(asOf?: string): Promise<Ladder> {
    return toCamelCase((await apiClient.get(`${base}/ladder`, { params: { as_of: asOf } })).data);
  },
  async run(params: { trade_date?: string; backfill_days?: number; missing_only?: boolean }): Promise<{ taskId: string }> {
    return toCamelCase((await apiClient.post(`${base}/run`, params)).data);
  },
};
